xtra-libp2p-offer/src/current/protocol.rs

```diff
@@ -0,0 +1,178 @@
+use asynchronous_codec::FramedRead;
+use asynchronous_codec::FramedWrite;
+use asynchronous_codec::JsonCodec;
+use asynchronous_codec::JsonCodecError;
+use futures::AsyncReadExt;
+use futures::AsyncWriteExt;
+use futures::SinkExt;
+use futures::StreamExt;
+use model::olivia::BitMexPriceEventId;
+use model::ContractSymbol;
+use model::FundingRate;
+use model::Leverage;
+use model::OfferId;
+use model::OpeningFee;
+use model::Position;
+use model::Price;
+use model::Timestamp;
+use model::TxFeeRate;
+use model::Usd;
+use serde::Deserialize;
+use serde::Serialize;
+use std::fmt;
+use time::Duration;
+
+pub(crate) async fn send<S>(sink: S, offers: Offers) -> Result<(), JsonCodecError>
+where
+    S: AsyncWriteExt + Unpin,
+{
+    tracing::info!(?offers, "Sending offers");
+
+    let mut framed = FramedWrite::new(sink, JsonCodec::<Offers, ()>::new());
+    framed.send(offers).await?;
+    MESSAGES_SENT.inc();
+
+    Ok(())
+}
+
+pub(crate) async fn recv<S>(stream: S) -> Result<Offers, ReceiveError>
+where
+    S: AsyncReadExt + Unpin,
+{
+    let mut framed = FramedRead::new(stream, JsonCodec::<(), Offers>::new());
+
+    let offers = framed.next().await.ok_or(ReceiveError::Terminated)??;
+
+    tracing::info!(?offers, "Received offers");
+
+    MESSAGES_RECEIVED.inc();
+
+    Ok(offers)
+}
+
+#[derive(Clone, Serialize, Deserialize, PartialEq, Debug)]
+pub(crate) struct Offers(Vec<Offer>);
```

Comment:
🛠 After another look I am not sure I agree with this model.

This model *implies* that the maker sends sane data.
A maker could send multiple offers for the same `contract_symbol` and `position` with this API.
Can't we be more explicit here?

This is what popped into my head:

* long and short are fixed
* the contract_symbold should scale, i.e. we might have more in the future

This leads me to a model like this:

```rust
pub struct MakerOffers {
   long: HashMap<ContractSymbol, Offer>,
   short: HashMap<ContractSymbol, Offer>
}
```

Additionally I would add a constructor that makes it explicit that one always has to set them for both symbols.
Alternatively we could also go for explicit setters, but I think the constructor it better to ensure we always handle the `Option`.

And getters for the respective one. 

```rust
impl MakerOffers {
    pub new(
        btcusd_long: Option<Offer>, 
        btcusd_short:Option<Offer>, 
        ethusd_long:Option<Offer>, 
        ethusd_short: Option<Offer>
    ) -> Self {
        if let Some(btcusd_long) = btcusd_long {
            self.long.insert(ContractSymbol::BtcUsd, btcusd_long);
        }

        ...
    }
}
```

This will scale as good over the wire, but is more explicit about the data. We can roll with this simpler model of `Vec` if you think it's not worth it, but I think it would be better to be explicit.

---

It is then questionable of the `Offer` should have a position; but I would be OK with that redundancy. The API to set long/short within `Offers` can ensure that long is always set to `long` internally.

---

If you find the `HashMap` too cumbersome we can still model the internals using a `Vec` but I think we should still restrict what can be set in that `Vec`. At the moment the code allows too arbitrary things, because we map from one `Vec` into another, I don't think that's good design.

Reply:
> 🤔 After another look I am not sure I agree with this model.
> 
> This model _implies_ that the maker sends sane data. A maker could send multiple offers for the same `contract_symbol` and `position` with this API. Can't we be more explicit here?
> 
> This is what popped into my head:
> 
>     * long and short are fixed
> 
>     * the contract_symbold should scale, i.e. we might have more in the future
> 
> 
> This leads me to a model like this:
> 
> ```rust
> pub struct MakerOffers {
>    long: HashMap<ContractSymbol, Offer>,
>    short: HashMap<ContractSymbol, Offer>
> }
> ```

I had something similar to this at some point, but I decided against it because you don't actually get that much for the price of making the message more complicated. The taker still has to make sense of the offers they just received together with the offers they already knew about. It's not any more complicated to iterate through a simple list applying the same logic. In the end the taker will always have to:

1. Combine offers received and offers stored.
2. Only keep one copy per `(ContracSymbol, Position)`, always choosing the newest one.

> It is then questionable of the `Offer` should have a position; but I would be OK with that redundancy. The API to set long/short within `Offers` can ensure that long is always set to `long` internally.

I would only opt for something like what you suggest if the `Offer` type sent over the wire omits both the position and the contract symbol. Otherwise the taker would still have to check that offers are correctly categorised on reception. And we would have to maybe ignore offers that are incorrectly categorised?

> Additionally I would add a constructor that makes it explicit that one always has to set them for both symbols. 

I don't see the need for this. Particularly not at the protocol level. Why is it important to limit what offers can be sent when? I expect the maker to want to send offers for only one contract symbol most of the time.

> If you find the `HashMap` too cumbersome we can still model the internals using a `Vec` but I think we should still restrict what can be set in that `Vec`. At the moment the code allows too arbitrary things, because we map from one `Vec` into another, I think at least the model should ensure sanity by e.g. a constructor that restricts the combinations to what is possible.

I am not fully seeing your concern here. Whether the maker sends 4 new offers for all the supported combinations or, for example, 20 copies of the same offer, the taker will still have to process them in the same way: keep the newest per `(ContractSymbol, Position)` pair.

Reply:
> I am not fully seeing your concern here. Whether the maker sends 4 new offers for all the supported combinations or, for example, 20 copies of the same offer, the taker will still have to process them in the same way: keep the newest per (ContractSymbol, Position) pair.

As long as the `maker_creation_timestamp` is sane and allow sorting we will be able to pick a "newest", otherwise it will get tricky. The taker will have an error-path to deal with that is: I received an offer with different values for the same symbol + position + timestamp - which one shall the taker pick?
 
But fair enough, this is somewhat fabricated.  I think you might be right that the cost of modelling it more complicated it not worth it. This model will scale and that's what's important.